### PR TITLE
mgmt: mcumgr: add user data to mgmt handler

### PR DESCRIPTION
**Pull request by kunoh**
SMP: Add user data to SMP callback handler

User data can be passed on to the handlers.
Currently handlers can only get access to user data by using global variables.
Existing SMP commands are tested on a NXP RT1060 board using mcumgr

RFC: https://github.com/zephyrproject-rtos/zephyr/issues/60616

Signed-off-by: Kuno Heltborg <kunoh@live.dk>

### CODE REVIEW

File: include/zephyr/mgmt/mcumgr/mgmt/mgmt.h
Comment (kartben) [inline include/zephyr/mgmt/mcumgr/mgmt/mgmt.h:177]: Add Doxygen comment for the new parameter

**Review by kartben — CHANGES_REQUESTED** [review]
Thanks for the PR!
This probably also needs an entry added in the release notes, as people's existing handlers will require updating.

**Review by nordicjm — CHANGES_REQUESTED** [review]
A few comments:

- What do you need this for?
- This is a stable API and a breaking change, so you would need to follow https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes
- There are handlers that have not been converted

File: include/zephyr/mgmt/mcumgr/mgmt/mgmt.h
Comment (de-nordic) [inline include/zephyr/mgmt/mcumgr/mgmt/mgmt.h:185]: I do not really like the addition. The PR does not provide reason for the change and addition. So far we have been able to live without this, while extending support for different commands.

**Review by de-nordic — CHANGES_REQUESTED** [review]
Lack rationale for the addition.

**Comment by kunoh** [comment]
RFC: [#60616](https://github.com/zephyrproject-rtos/zephyr/issues/60616)

**Comment by kunoh** [comment]
> A few comments:
> 
> * What do you need this for?
> * This is a stable API and a breaking change, so you would need to follow https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes
> * There are handlers that have not been converted

Thank you!

* A handler can currently not reach dynamically allocated variables, which I need in order for it to reach a member variable within a class, that is dynamically allocated. I would like not to have global instantiated classes in order for handlers to access it's members. By passing user data I can pass a pointer to the instance of the class.
* I have opened a RFC.
* Should I update missing handlers, or maybe it's better to wait for a final decision on this?

File: include/zephyr/mgmt/mcumgr/mgmt/mgmt.h
Comment (kunoh) [reply include/zephyr/mgmt/mcumgr/mgmt/mgmt.h:185]: Hi @de-nordic, I just added a RFC, please check the comment below

File: include/zephyr/mgmt/mcumgr/mgmt/mgmt.h
Comment (kunoh) [reply include/zephyr/mgmt/mcumgr/mgmt/mgmt.h:177]: Thank you! Will update the PR if the RFC is approved :)

**Pull request closed without merging**
